snuba/clickhouse/http.py

```diff
@@ -73,15 +94,80 @@ def write(self, rows: Iterable[WriterTableRow]):
             chunked=True,
         )
 
-        if response.status != 200:
-            # XXX: This should be switched to just parse the JSON body after
-            # https://github.com/yandex/ClickHouse/issues/6272 is available.
-            content = response.data.decode("utf8")
-            details = CLICKHOUSE_ERROR_RE.match(content)
-            if details is not None:
-                code, type, message = details.groups()
-                raise ClickhouseError(int(code), message)
-            else:
-                raise HTTPError(
-                    f"Received unexpected {response.status} response: {content}"
-                )
+        raise_for_error_response(response)
+
+
+def parse_and_transform_date(value: str) -> str:
+    return transform_date(dateutil_parse(value))
+
+
+def parse_and_transform_datetime(value: str) -> str:
+    return transform_datetime(dateutil_parse(value))
+
+
+transform_column_types = build_result_transformer(
+    [
+        (re.compile(r"^Date(\(.+\))?$"), parse_and_transform_date),
+        (re.compile(r"^DateTime(\(.+\))?$"), parse_and_transform_datetime),
+    ]
+)
+
+
+class HTTPReader(Reader[ClickhouseQuery]):
+    def __init__(
+        self, host: str, port: int, settings: Optional[Mapping[str, str]] = None
+    ):
+        if settings is not None:
+            assert "query_id" not in settings, "query_id cannot be passed as a setting"
+
```

Comment:
Sort of.

The settings passed here are supposed to be closer to "session settings" (`client_settings` for the `ClickhousePool` constructor) than they are to "query settings", but we don't use [ClickHouse HTTP sessions](https://www.altinity.com/blog/2018/9/28/progress-reports-for-long-running-queries-via-http-protocol) here due to not wanting to deal with session timeouts, etc. I needed to be some way to allow constant (session-like) settings here to support things like `output_format_json_quote_64bit_integers` without leaking that into `raw_query`. This is more of a manifestation of `query_id` being passed as a setting, rather than some other way (e.g. a header.) I don't know how the native driver here would deal with `query_id`, such as whether or not it considers that a valid setting, or if that's an implementation detail of the way settings are implemented in the HTTP interface. (I don't know what it would do if `query_id` was provided in both the `settings`  mapping as well as the `query_id` keyword argument to [`execute`](https://clickhouse-driver.readthedocs.io/en/latest/api.html#clickhouse_driver.Client.execute).)